tutorials/arvo/ames.md

```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+the vane as a `%done` gift. Naxplanations may only be acked, never nacked.
+
+#### (N)ack packets
+
+The contents of a given (n)ack packet are deterministic. The `content` (i.e. the
+encrypted portion of the packet) of a fragment ack is obtained by encrypting the `+jam` of the
+following noun:
+
+```hoon
+[our-life her-life bone message-num fragment-num]
+```
+Each datum in this noun is an atom with the aura `@ud` or an aura that nests
+under `@ud`.
+
+Here, `our-life` refers to the [`life`](@/docs/glossary/breach), or revision
+number, of the acking ship's networking keys, and `her-life` is the `life` of
+the ack-receiving ship's networking keys. `bone` is an opaque number identifying the flow, `message-num` denotes the number of the
+message in the flow identified by `bone`, and
+`fragment-num` denotes the number of the fragment of the message identified by
+`message-num` being acked. 
+
+A message (n)ack is an extended fragment ack that is obtained by encrypting the
+`+jam` of the following noun:
+
+```hoon
+[our-life her-life bone message-num fragment-num ok=? lag=@dr]
+```
+`ok` is a flag that is set to `%.y` for a message ack and `%.n` for a message
+nack. In the future, `lag` will be used to
+describe the time it took to process a message, but for now it is always zero.
+
+
+### Flows
+
+Flows are asymmetric communication channels along which two ships send and
+receive packets, and all Ames packets are part of some flow.
+
+Flows are asymmetric in that one ship is always the sender of
+`%plea`s and the other is always the sender of `%boon`s on a given flow. Thus ships will
+typically have multiple flows between them.
+
+The sequence of `%plea`s in a flow is totally ordered, though
+the packets which make them up need not be. `%boon`s are totally ordered in the
+same fashion, but there is not a coordinated ordering between `%plea`s and
+`%boon`s in a given flow beyond the implicit one arising from the fact that a
+`%boon` is always a response to a `%plea`.
+
+Inside of Ames, each flow has four sequential opaque `@ud`s
+called bones that are unique to that flow. Thus the flow itself is often
+referred to by its first bone. The set of types of packets is partitioned into
+four subsets, each of which is assigned a bone. Each bone is a one-way street
+for packets to travel along, so e.g. acks to packets making up a `%plea` are
+sent along different bones.
+
+We give an example of such a partition. Let flow 12 be a flow between `~bacbel-tagfeg` and `~worwel-sipnum` with
+`~bacbel-tagfeg` as the `%plea` sender. Then bones 12-15 are associated with the
+flow, and the types of packets for each bone are:
+
+ - bone 12, `%plea`s and acks to `~worwel-sipnum`,
+ - bone 13, `%boon`s and (n)acks to `~bacbel-tagfeg`,
+ - bone 14, acks of naxplanations to `~worwel-sipnum`,
+ - bone 15, naxplanations to `~bacbel-tagfeg`.
+
+Each bone is handled separately by congestion control, and this is the reason
```

Comment:
Separate congestion control is one reason to separate bones.  There are two others.

It's important that I be able to start a flow with you without risk of race conditions.  That is, if I try to start a flow with you at the same time you try to start a flow with me, when we receive each other's packets, that shouldn't cause a conflict.  Flipping the last bit of the bone based on whether a packet is an ack or a message fragment allows for me to create flow 8 with you without that ever coming into conflict with the flow 8 that you created with me.

The reason to have four bones per flow, not just two, is that naxplanation packets shouldn't be able to conflict with happy-path packets.  Since naxplanations are generated dynamically in response to hearing messages, if they were  on the same bones as happy-path messages, they would have to occupy new numbers within those flows.  Maybe it would be possible to get that to work, but it was easier to be confident in the correctness of the design with naxplanations on their own bones that can't conflict with the happy path.

Reply:
https://github.com/urbit/docs/pull/957/commits/d42f0a3aee11fd2b410448332da010a1a880ac23#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR302-R320

How's this?

Reply:
Whoops, somehow started a new review for that comment. I'm gonna leave it just to ensure I don't fuck something else up.

Reply:
LGTM